### PR TITLE
Search for projects/challenges by tag(s)

### LINKED ISSUE #15: Filter by tag
We support searching by challenge/project name and description, but filtering by tags is lacking. Clicking on a tag should show only challenges/projects that have it.

### PR DESCRIPTION
**Pull request by dulldesk**
Users can filter projects **or** challenges by tags. 

Implements #15

### CODE REVIEW

File: app/templates/explore.html [hunk 1]
```diff
@@ -2,7 +2,37 @@
 {% load static %}
 
 {% block head %}
-	<title>Explore | Stratus</title>
+  <title>Explore | Stratus</title>
+  <link rel="stylesheet" href="{% static 'vendor/chosen/chosen.min.css' %}">
+  <script src="{% static 'vendor/chosen/chosen.jquery.min.js' %}"></script>
+  <style>
```
Comment (fluix-dev) [inline app/templates/explore.html]: This seems like a lot of styles, let's put this in our stylesheet.

**Comment by dulldesk** [comment]
Some remarks:
- The CSS used by Chosen is inconsistent with that of Cirrus (and the entire site). While 79e9d51 attempts to rectify this, inconsistencies still exist, especially for different viewport sizes. 
- Having two "search bars" (one for querying projects/challenges, one for filtering tags) looks odd. Any suggestions to change this?
- Is the Chosen license required?

File: app/views.py [hunk 2]
```diff
@@ -19,19 +19,45 @@ class IndexView(TemplateView):
...
-            )
+        else:
+            initiative = self.request.GET["type"]
+            context["type"] = [initiative]
+
+            if initiative == "challenge":
+                queryset = Challenge.objects.all()
+
+            elif initiative == "project":
+                queryset = Project.objects.all()
```
Comment (fluix-dev) [inline app/views.py]: We need to do something if it's neither of these.

File: app/views.py [hunk 3]
```diff
@@ -19,19 +19,45 @@ class IndexView(TemplateView):
...
-            context["projects"] = Project.objects.all().filter(
-                Q(name__contains=query) | Q(description__contains=query)
-            )
+        else:
+            initiative = self.request.GET["type"]
+            context["type"] = [initiative]
+
+            if initiative == "challenge":
+                queryset = Challenge.objects.all()
+
```
Comment (fluix-dev) [inline app/views.py]:
```suggestion
```

File: app/views.py [hunk 4]
```diff
@@ -19,19 +19,45 @@ class IndexView(TemplateView):
...
+            if "q" in self.request.GET:
+                query = self.request.GET["q"]
+                context["q"] = query
+
+                queryset = queryset.filter(
+                    Q(name__icontains=query) | Q(description__icontains=query)
+                )
+
+            context[initiative + "s"] = queryset.distinct()
```
Comment (fluix-dev) [inline app/views.py]: This is alongside HTML changes, the template for projects and challenges seems extremely similar. Can we not just call this `objects` and `{% for object in objects %}`.

**Review by fluix-dev — CHANGES_REQUESTED** [review]

**Comment by fluix-dev** [comment]
> Is the Chosen license required?

It's probably enough to have it inside the .css and .js files, but because we have a folder (`chosen`) anyways, let's leave the full license text.

**Comment by fluix-dev** [comment]
> Having two "search bars" (one for querying projects/challenges, one for filtering tags) looks odd. Any suggestions to change this?

I haven't seen it, but I can imagine it's not pretty. I think the better solution is similar GitHub topics where you don't filter by more than one tag, but if you and the others want multiple tag filtering, I won't reject it.

**Edit**: This also removes the Chosen dependency.

**Edit**: Others seem to want it.

File: app/views.py (hunk 2 above)
Comment (dulldesk) [reply app/views.py]: Perhaps we could just ignore it and resort to displaying the unqueried `explore` page (i.e. when checking if a `tag` parameter exists, also verify its value to be either `"project"` or `"challenge"`)?

Alternative could be an error page

Comment (fluix-dev) [reply app/views.py]: Error page is too much work, just assume it's absent.

Comment (fluix-dev) [reply app/views.py]: So either redirect to unqueried or return unqueried.

File: app/views.py (hunk 4 above)
Comment (dulldesk) [reply app/views.py]: Do you refer to these lines? (If so, that works, will implement)

https://github.com/KuriusMTL/stratus/pull/27/commits/6335f38f581698274814e70c20d8e85e2145e9b3#diff-c80b5e347b555901a695e72476d025b0R50-R110

Comment (dulldesk) [reply app/views.py]: (https://github.com/KuriusMTL/stratus/pull/27/commits/9ed00ea804203bffaf6a1bb9d52c3a0f9ac74a6f)

File: app/views.py (hunk 2 above)
Comment (dulldesk) [reply app/views.py]: (8ba3a29846c08b2f852257a477b0426fa98f9deb)

File: app/views.py [hunk 5]
```diff
@@ -11,27 +11,56 @@
 from django.views.generic.edit import FormView, UpdateView
 
 from django.contrib.contenttypes.models import ContentType
```
Comment (fluix-dev) [inline app/views.py]: I know you didn't write this line, but you can merge all of these with the above imports (everything that *doesn't* start with `app.`), thanks.

File: app/views.py [hunk 6]
```diff
@@ -11,27 +11,56 @@
...
+        context["tags"] = Tag.objects.all()
+        context["selected_tags"] = []
+
+        if "type" not in self.request.GET or (
+            "q" not in self.request.GET and "tag" not in self.request.GET
+        ):
+            context["objects"] = {
+                "challenge": Challenge.objects.all()[:3],
+                "project": Project.objects.all()[:9],
+            }
```
Comment (fluix-dev) [inline app/views.py:35]: `return` immediately so you can remove the indentation that comes with the `else`.

File: app/views.py [hunk 7]
```diff
@@ -11,27 +11,56 @@
...
+        else:
+            initiative = self.request.GET["type"]
+
+            if initiative == "challenge":
+                queryset = Challenge.objects.all()
+            elif initiative == "project":
+                queryset = Project.objects.all()
+
+            if "tag" in self.request.GET:
+                query = self.request.GET.getlist("tag")
```
Comment (fluix-dev) [inline app/views.py]: [`getlist`](https://docs.djangoproject.com/en/3.1/ref/request-response/#django.http.QueryDict.getlist) will return an empty list by default, so you can omit the if statement above.

File: app/templates/explore.html [hunk 8]
```diff
@@ -18,91 +49,80 @@ <h3 class="sub-title faded">Conquer the world's challenges - together</h3>
...
+  {% for name, object_list in objects.items %}
+    {# searching options #}
+    <form action=".">
+      <input type="hidden" name="type" value="{{ name }}">
+      <div class="row">
+        <h1 class="header-margin">{{ name|title }}s</h1>
+        <div class="col-12">
+          <div class="input-control">
+            <input type="text" name="q" class="input-contains-icon" placeholder="{{ name|title }} Search"
+              {% if q %} value="{{ q }}" {% endif %} >
```
Comment (fluix-dev) [inline app/templates/explore.html]:
```suggestion
              value="{{ q|default:"" }}">
```

File: app/views.py [hunk 9]
```diff
@@ -11,27 +11,56 @@
...
+
+                for tag in query:
+                    currtags = Tag.objects.filter(name__iexact=tag)
+                    queryset = queryset.filter(Q(tags__in=currtags))
+                    context["selected_tags"] += currtags
+
+            if "q" in self.request.GET:
+                query = self.request.GET["q"]
+                context["q"] = query
```
Comment (fluix-dev) [inline app/views.py]:
```suggestion
                query = context["q"] = self.request.GET["q"]
```

File: app/views.py [hunk 10]
```diff
@@ -11,27 +11,56 @@
...
+            elif initiative == "project":
+                queryset = Project.objects.all()
+
+            if "tag" in self.request.GET:
+                query = self.request.GET.getlist("tag")
+
+                for tag in query:
+                    currtags = Tag.objects.filter(name__iexact=tag)
+                    queryset = queryset.filter(Q(tags__in=currtags))
+                    context["selected_tags"] += currtags
```
Comment (fluix-dev) [inline app/views.py]: This runs a *lot* of queries, you should assembly a list of `pk`s for each of the valid tags and then only call `queryset.filter` once.

**Review by fluix-dev — CHANGES_REQUESTED** [review]

File: app/views.py (hunk 7 above)
Comment (fluix-dev) [reply app/views.py]: Also probably name this `tags` and not `query`.

File: app/views.py [hunk 11]
```diff
@@ -1,37 +1,57 @@
...
+        return super(IndexView, self).dispatch(request, *args, **kwargs)
+
     def get_context_data(self, **kwargs):
         context = super().get_context_data(**kwargs)
-        if "type" not in self.request.GET or "q" not in self.request.GET:
-            context["challenges"] = Challenge.objects.all()[:3]
-            context["projects"] = Project.objects.all()[:9]
+
+        context["tags"] = Tag.objects.all()
+        context["selected_tags"] = []
```
Comment (fluix-dev) [inline app/views.py]: I think this isn't necessary.
```suggestion

```

File: app/views.py [hunk 12]
```diff
@@ -1,37 +1,57 @@
...
-            context["challenges"] = Challenge.objects.all().filter(
-                Q(name__contains=query) | Q(description__contains=query)
-            )
-        elif self.request.GET["type"] == "project":
-            context["projects"] = Project.objects.all().filter(
-                Q(name__contains=query) | Q(description__contains=query)
-            )
+
+        initiative = self.request.GET["type"]
+
```
Comment (fluix-dev) [inline app/views.py]:
```suggestion

```

File: app/views.py [hunk 13]
```diff
@@ -1,37 +1,57 @@
...
+        if initiative == "challenge":
+            queryset = Challenge.objects.all()
+        elif initiative == "project":
+            queryset = Project.objects.all()
+
+        context["selected_tags"] = Tag.objects.filter(name__in=self.request.GET.getlist("tag"))
+
```
Comment (fluix-dev) [inline app/views.py]:
```suggestion

```

File: app/views.py [hunk 14]
```diff
@@ -1,37 +1,57 @@
...
-        if "type" not in self.request.GET or "q" not in self.request.GET:
-            context["challenges"] = Challenge.objects.all()[:3]
-            context["projects"] = Project.objects.all()[:9]
+
+        context["tags"] = Tag.objects.all()
+        context["selected_tags"] = []
+
+        if "type" not in self.request.GET or (
+            "q" not in self.request.GET and "tag" not in self.request.GET
+        ):
```
Comment (fluix-dev) [inline app/views.py:31]: This doesn't look great but I can't think of anything better.

**Review by fluix-dev — APPROVED** [review]
Last few nits, otherwise LGTM. Thanks so much!

File: app/views.py (hunk 14 above)
Comment (dulldesk) [reply app/views.py:31]:
```python
        if not self.request.GET.get("type") or (
            not self.request.GET.get("q") and not self.request.GET.get("tag")
        ):
```

Is this preferable by any means?

**Pull request merged**